### ISSUE TITLE
[parallel-queries] Refactor error message handling to better integrate with parallel queries

### ISSUE THREAD

**Issue by michaelwoerister** [rust-lang/rust#49737]
At the moment error message handling uses a [callback-based setup](https://github.com/rust-lang/rust/blob/2253216ec500601f9c2998196cb1b0a1ed0ae576/src/librustc_errors/lib.rs#L285-L286) in order to handle diagnostics tracking during query evaluation. This could probably be handled in a cleaner way.

Also, the current system has no way of making error message output somewhat deterministic under parallel execution.

Both problems should be solved by re-factoring the system to better fit the changed needs.

**Comment by sasurau4** [comment]
It seems that this issue solved with https://github.com/rust-lang/rust/pull/67614.

**Comment by michaelwoerister** [comment]
I don't think that PR takes care of emitting error messages in a deterministic order, right?

**Comment by Zoxc** [comment]
I'm not sure we'd want to do this is it would require error messages to be delayed.
